internal/controller/nodetemplate_controller.go

```diff
@@ -0,0 +1,454 @@
+				"node", candidate.Name,
+				"nodetemplate", nodeTemplate.Name,
+				"phase", "ensuring",
+			)
+			continue
+		}
+		err := r.Client.Update(ctx, &candidate)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		log.Info("allocated node",
+			"node", candidate.Name,
+			"nodetemplate", nodeTemplate.Name,
+		)
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *NodeTemplateReconciler) reflectNodeConfiguration(node *corev1.Node, nodeTemplate *nyallocatorv1.NodeTemplate) {
+	if nodeTemplate.Spec.Template.Metadata.Labels != nil && node.Labels == nil {
+		node.Labels = make(map[string]string)
+	}
+	if nodeTemplate.Spec.Template.Metadata.Annotations != nil && node.Annotations == nil {
+		node.Annotations = make(map[string]string)
+	}
+	// remove spare taint and label, and add node template reference label
+	for i, taint := range node.Spec.Taints {
+		if taint.Key == SpareTaintKey {
+			node.Spec.Taints = slices.Delete(node.Spec.Taints, i, i+1)
+		}
+	}
+	delete(node.Labels, SpareRoleLabelKey)
+	node.Labels[NodeTemplateReferenceLabelKey] = nodeTemplate.Name
+
+L:
+	for _, taint := range nodeTemplate.Spec.Template.Spec.Taints {
+		for i, t := range node.Spec.Taints {
+			if t.Key == taint.Key {
+				node.Spec.Taints[i] = taint
+				continue L
+			}
+		}
+		node.Spec.Taints = append(node.Spec.Taints, taint)
+	}
+
+	node.Labels["node-role.kubernetes.io"+"/"+nodeTemplate.Name] = "true"
+	for k, v := range nodeTemplate.Spec.Template.Metadata.Labels {
+		node.Labels[k] = v
+	}
+
+	for k, v := range nodeTemplate.Spec.Template.Metadata.Annotations {
+		node.Annotations[k] = v
+	}
+}
+
+func selectNodes(spareNodes map[string]corev1.Node, zoneCount map[string]int) corev1.Node {
+	maxCountPerZone := 100
+	maxScore := -1
+	maxScoreNode := ""
+	for k, n := range spareNodes {
+		zoneCountScore := 0
+		if zone, ok := n.Labels[ZoneLabelKey]; ok {
+			zoneCountScore = maxCountPerZone - zoneCount[zone]
+		}
+		score := zoneCountScore * 10
+		if score > maxScore {
+			maxScore = score
+			maxScoreNode = k
+		}
+	}
+	return spareNodes[maxScoreNode]
+}
+
+func (r *NodeTemplateReconciler) updateStatus(ctx context.Context, nodeTemplate *nyallocatorv1.NodeTemplate, reason *InsufficientReason, reconcileError error) error {
+	nodes := corev1.NodeList{}
+	err := r.Client.List(ctx, &nodes, client.MatchingLabels{NodeTemplateReferenceLabelKey: nodeTemplate.Name})
+	if err != nil {
+		return err
+	}
+	selector, err := metav1.LabelSelectorAsSelector(nodeTemplate.Spec.Selector)
+	if err != nil {
+		return err
+	}
+	underManagementNodes := []corev1.Node{}
+	for _, node := range nodes.Items {
+		if selector.Matches(labels.Set(node.Labels)) {
+			underManagementNodes = append(underManagementNodes, node)
+		}
+	}
+
+	nodeTemplate.Status.ReconcileInfo.ObservedGeneration = nodeTemplate.Generation
+
+	nodeTemplate.Status.CurrentNodes = len(underManagementNodes)
+
+	conditionSufficient := metav1.Condition{
+		Type:               ConditionSufficient,
+		LastTransitionTime: metav1.Now(),
```

Comment:
Only change the time when the sufficient status changes.

Reply:
I'll use `meta.SetStatusCondition` instead.